Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+            var result = {};
+            result[queryString] = undefined;
+            return result;
+        }
+
+        return queryToObject(queryString);
+    }
+
+    /**
+     * @private
+     */
+    function stringifyQuery(queryObject) {
+        var keys = Object.keys(queryObject);
+
+        // We have 1 key with an undefined value, so this is just a string, not key/value pairs
+        if (keys.length === 1 && !defined(queryObject[keys[0]])) {
+            return keys[0];
+        }
+
+        return objectToQuery(queryObject);
+    }
+
+    /**
+     * @private
+     */
+    function encodeValues(object) {
+        var result = {};
+        for(var key in object) {
+            if (object.hasOwnProperty(key)) {
+                result[key] = encodeURIComponent(object[key]);
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * A resource the location and any other parameters we need to retrieve it or create derived resources.
+     *
+     * @param {Object} options An object with the following properties
+     * @param {String} options.url The url of the resource.
+     * @param {Object} [options.queryParameters] An object containing query parameters that will be sent when retrieving the resource.
+     * @param {Object} [options.templateValues] Key/Value pairs that are used to replace template values (eg. {x}).
+     * @param {Object} [options.headers={}] Additional HTTP headers that will be sent.
+     * @param {String} [options.responseType] The type of response.
+     * @param {String} [options.method='GET'] The method to use.
+     * @param {Object} [options.data] Data that is sent with the resource request.
+     * @param {String} [options.overrideMimeType] Overrides the MIME type returned by the server.
+     * @param {DefaultProxy} [options.proxy] A proxy to be used when loading the resource.
+     * @param {Boolean} [options.isDirectory=false] The url should be a directory, so make sure there is a trailing slash.
+     * @param {Function} [options.retryCallback] The Function to call when a request for this resource fails. If it returns true, the request will be retried.
+     * @param {Number} [options.retryAttempts=0] The number of times the retryCallback should be called before giving up.
+     * @param {Request} [options.request] A Request object that will be used. Intended for internal use only.
+     *
+     * @constructor
+     */
+    function Resource(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.string('options.url', options.url);
+        //>>includeEnd('debug');
+
+        this._url = '';
+        this._templateValues = encodeValues(defaultValue(options.templateValues, {}));
+        this._queryParameters = defaultValue(options.queryParameters, {});
+        this._isDirectory = defaultValue(options.isDirectory, false);
+        this.url = options.url;
```

Comment:
Probably better to have this be the last line of the constructor, that way the class is completely initialized with all values first.